packages/liquidator/index.js

```diff
@@ -53,24 +61,52 @@ async function run({
     const getTime = () => Math.round(new Date().getTime() / 1000);
 
     // Load unlocked web3 accounts and get the networkId.
-    const [accounts, networkId] = await Promise.all([web3.eth.getAccounts(), web3.eth.net.getId()]);
+    const [detectedContract, accounts, networkId] = await Promise.all([
+      findContractVersion(empAddress, web3),
+      web3.eth.getAccounts(),
+      web3.eth.net.getId()
+    ]);
+
+    // Append the contract version and type to the liquidatorConfig, if the liquidatorConfig does not already contain one.
+    if (!liquidatorConfig) liquidatorConfig = {};
+    if (!liquidatorConfig.contractVersion) liquidatorConfig.contractVersion = detectedContract.contractVersion;
+    if (!liquidatorConfig.contractType) liquidatorConfig.contractType = detectedContract.contractType;
+
+    // Check that the version and type is supported. Note if either is null this check will also catch it.
+    if (
+      SUPPORTED_CONTRACT_VERSIONS.filter(
+        vo => vo.contractType == liquidatorConfig.contractType && vo.contractVersion == liquidatorConfig.contractVersion
+      ).length == 0
+    )
+      throw new Error(
+        `Contract version specified or inferred is not supported by this bot. Loaded/inferred contractVersion:${
+          liquidatorConfig.contractVersion
+        } & contractType:${liquidatorConfig.contractType} is not part of ${JSON.stringify(SUPPORTED_CONTRACT_VERSIONS)}`
+      );
 
-    // Setup contract instances.
-    const voting = new web3.eth.Contract(getAbi("Voting"), getAddress("Voting", networkId));
-    const emp = new web3.eth.Contract(getAbi("ExpiringMultiParty"), empAddress);
+    // Setup contract instances. This uses the contract version pulled in from previous step. Voting is hardcoded to latest main net version.
+    const voting = new web3.eth.Contract(getAbi("Voting", "1.2.2"), getAddress("Voting", networkId));
+    const emp = new web3.eth.Contract(
+      getAbi(liquidatorConfig.contractType, liquidatorConfig.contractVersion),
+      empAddress
+    );
 
     // Returns whether the EMP has expired yet
-    const checkIsExpiredPromise = async () => {
-      const [expirationTimestamp, contractTimestamp] = await Promise.all([
-        emp.methods.expirationTimestamp().call(),
+    const checkIsExpiredOrShutdownPromise = async () => {
+      const [expirationOrShutdownTimestamp, contractTimestamp] = await Promise.all([
+        liquidatorConfig.contractType == "ExpiringMultiParty"
+          ? emp.methods.expirationTimestamp().call()
+          : emp.methods.emergencyShutdownTimestamp().call(),
         emp.methods.getCurrentTime().call()
       ]);
       // Check if EMP is expired.
-      if (Number(contractTimestamp) >= Number(expirationTimestamp)) {
+      if (Number(contractTimestamp) >= Number(expirationOrShutdownTimestamp) && expirationOrShutdownTimestamp > 0) {
```

Comment:
Should the RHS of the `&&` be `Number(expirationOrShutdownTimestamp)` since we do that on the LHS?

Reply:
ye, good point. done